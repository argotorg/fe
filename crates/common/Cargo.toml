[package]
name = "fe-common"
version = "0.12.0-alpha"
authors = ["The Fe Developers <snakecharmers@ethereum.org>"]
edition = "2021"
license = "Apache-2.0"
repository = "https://github.com/ethereum/fe"

[dependencies]
fe-library = {path = "../library", version = "^0.12.0-alpha"}
tiny-keccak = { version = "2.0", features = ["keccak"] }
hex = "0.4"
codespan-reporting = "0.11.1"
serde = { version = "1", features = ["derive"] }
ron = "0.5.1"
difference = "2.0"
num-traits = "0.2.14"
once_cell = "1.8.0"
include_dir = "0.6.0"
indexmap = "1.6.2"
smol_str = "0.1.21"
