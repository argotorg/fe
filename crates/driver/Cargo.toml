[package]
name = "fe-driver"
version = "0.13.0-alpha"
authors = ["The Fe Developers <snakecharmers@ethereum.org>"]
edition = "2021"
license = "GPL-3.0-or-later"
repository = "https://github.com/ethereum/fe"

[features]
# Includes the solidity compiler for Yul -> bytecode compilation.
solc-backend = ["fe-yulc"]

[dependencies]
serde_json = "1.0"
fe-abi = {path = "../abi", version = "^0.13.0-alpha"}
fe-analyzer = {path = "../analyzer", version = "^0.13.0-alpha"}
fe-common = {path = "../common", version = "^0.13.0-alpha"}
fe-lowering = {path = "../lowering", version = "^0.13.0-alpha"}
fe-mir = {path = "../mir", version = "^0.13.0-alpha"}
fe-parser = {path = "../parser", version = "^0.13.0-alpha"}
fe-yulgen = {path = "../yulgen", version = "^0.13.0-alpha"}
fe-yulc = {path = "../yulc", version = "^0.13.0-alpha", features = ["solc-backend"], optional = true}
indexmap = "1.6.2"
vfs = "0.5.1"
smol_str = "0.1.21"
toml = "0.5.8"
