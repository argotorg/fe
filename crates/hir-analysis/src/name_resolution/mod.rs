pub mod diagnostics;

mod import_resolver;
mod name_resolver;
mod path_resolver;
mod visibility_checker;

use either::Either;
use hir::{
    analysis_pass::ModuleAnalysisPass,
    diagnostics::DiagnosticVoucher,
    hir_def::{
        scope_graph::ScopeId, Expr, ExprId, GenericArgListId, IdentId, IngotId, ItemKind, Partial,
        Pat, PatId, PathId, TopLevelMod, TraitRefId, TypeId,
    },
    visitor::prelude::*,
};
pub use import_resolver::ResolvedImports;
pub use name_resolver::{
    NameDerivation, NameDomain, NameQuery, NameRes, NameResBucket, NameResKind, QueryDirective,
};
pub use path_resolver::EarlyResolvedPath;
use rustc_hash::FxHashSet;

use self::{
    diagnostics::{ImportResolutionDiagAccumulator, NameResDiag, NameResolutionDiagAccumulator},
    import_resolver::DefaultImporter,
    name_resolver::{NameResolutionError, ResolvedQueryCacheStore},
    path_resolver::EarlyPathResolver,
};
use crate::HirAnalysisDb;

// TODO: Implement `resolve_path` and `resolve_segments` after implementing the
// late path resolution.

/// Resolves the given path in the given scope.
/// It's not necessary to report any error even if the `EarlyResolvedPath`
/// contains some errors; it's always reported from [`PathAnalysisPass`].
pub fn resolve_path_early(
    db: &dyn HirAnalysisDb,
    path: PathId,
    scope: ScopeId,
) -> EarlyResolvedPath {
    resolve_segments_early(db, path.segments(db.as_hir_db()), scope)
}

/// Resolves the given path segments in the given scope.
/// It's not necessary to report any error even if the `EarlyResolvedPath`
/// contains some errors; it's always reported from [`PathAnalysisPass`].
pub fn resolve_segments_early(
    db: &dyn HirAnalysisDb,
    segments: &[Partial<IdentId>],
    scope: ScopeId,
) -> EarlyResolvedPath {
    // Obtain cache store for the given scope.
    let cache_store = resolve_path_early_impl(db, scope.top_mod(db.as_hir_db()));
    let importer = DefaultImporter;
    // We use the cache store that is returned from `resolve_path_early` to get
    // cached results immediately.
    let mut name_resolver = name_resolver::NameResolver::new_no_cache(db, &importer);

    let mut resolver = EarlyPathResolver::new(db, &mut name_resolver, cache_store);
    match resolver.resolve_segments(segments, scope) {
        Ok(res) => res.resolved,
        Err(_) => {
            // It's ok to ignore the errors here and returns an empty bucket because the
            // precise errors are reported from `PathAnalysisPass`.
            let bucket = NameResBucket::default();
            EarlyResolvedPath::Full(bucket)
        }
    }
}

/// Performs import resolution analysis. This pass only checks correctness of
/// the imports and doesn't emit other name resolutions errors.
pub struct ImportAnalysisPass<'db> {
    db: &'db dyn HirAnalysisDb,
}

impl<'db> ImportAnalysisPass<'db> {
    pub fn new(db: &'db dyn HirAnalysisDb) -> Self {
        Self { db }
    }

    pub fn resolve_imports(&self, ingot: IngotId) -> &'db ResolvedImports {
        resolve_imports(self.db, ingot)
    }
}

impl<'db> ModuleAnalysisPass for ImportAnalysisPass<'db> {
    fn run_on_module(&mut self, top_mod: TopLevelMod) -> Vec<Box<dyn DiagnosticVoucher>> {
        let ingot = top_mod.ingot(self.db.as_hir_db());
        resolve_imports::accumulated::<ImportResolutionDiagAccumulator>(self.db, ingot)
            .into_iter()
            .filter_map(|diag| (diag.top_mod(self.db) == top_mod).then(|| Box::new(diag) as _))
            .collect()
    }
}

/// Performs path resolution analysis. This pass checks all paths appeared in a
/// module for
/// - Existence
/// - Visibility
/// - Domain correctness
/// - Ambiguity
///
/// NOTE: This pass doesn't check the conflict of item definitions or import
/// errors. If you need to check them, please consider using
/// [`ImportAnalysisPass`] or [`DefConflictAnalysisPass`].
pub struct PathAnalysisPass<'db> {
    db: &'db dyn HirAnalysisDb,
}

impl<'db> PathAnalysisPass<'db> {
    pub fn new(db: &'db dyn HirAnalysisDb) -> Self {
        Self { db }
    }
}

impl<'db> ModuleAnalysisPass for PathAnalysisPass<'db> {
    fn run_on_module(&mut self, top_mod: TopLevelMod) -> Vec<Box<dyn DiagnosticVoucher>> {
        let errors =
            resolve_path_early_impl::accumulated::<NameResolutionDiagAccumulator>(self.db, top_mod);

        errors
            .into_iter()
            .filter_map(|err| {
                (!matches!(err, NameResDiag::Conflict(..))).then(|| Box::new(err) as _)
            })
            .collect()
    }
}

/// Performs conflict analysis. This pass checks the conflict of item
/// definitions.
pub struct DefConflictAnalysisPass<'db> {
    db: &'db dyn HirAnalysisDb,
}

impl<'db> DefConflictAnalysisPass<'db> {
    pub fn new(db: &'db dyn HirAnalysisDb) -> Self {
        Self { db }
    }
}

impl<'db> ModuleAnalysisPass for DefConflictAnalysisPass<'db> {
    fn run_on_module(&mut self, top_mod: TopLevelMod) -> Vec<Box<dyn DiagnosticVoucher>> {
        let errors =
            resolve_path_early_impl::accumulated::<NameResolutionDiagAccumulator>(self.db, top_mod);

        // TODO: `ImplCollector`.
        errors
            .into_iter()
            .filter_map(|err| matches!(err, NameResDiag::Conflict(..)).then(|| Box::new(err) as _))
            .collect()
    }
}

#[salsa::tracked(return_ref)]
pub(crate) fn resolve_imports(db: &dyn HirAnalysisDb, ingot: IngotId) -> ResolvedImports {
    let resolver = import_resolver::ImportResolver::new(db, ingot);
    let (imports, diags) = resolver.resolve_imports();
    for diag in diags {
        ImportResolutionDiagAccumulator::push(db, diag);
    }

    imports
}

/// Performs early path resolution and cache the resolutions for paths appeared
/// in the given module. Also checks the conflict of the item definitions.
///
/// NOTE: This method doesn't check
/// - the conflict in impl/impl-trait blocks since it requires ingot granularity
///   analysis.
/// - the path resolution errors at expression and statement level since it
///   generally requires type analysis
#[salsa::tracked(return_ref)]
pub(crate) fn resolve_path_early_impl(
    db: &dyn HirAnalysisDb,
    top_mod: TopLevelMod,
) -> ResolvedQueryCacheStore {
    let importer = DefaultImporter;
    let mut visitor = EarlyPathVisitor::new(db, &importer);

    let mut ctxt = VisitorCtxt::with_item(db.as_hir_db(), top_mod.into());
    visitor.visit_item(&mut ctxt, top_mod.into());

    for diag in visitor.diags {
        NameResolutionDiagAccumulator::push(db, diag);
    }
    visitor.inner.into_cache_store()
}

struct EarlyPathVisitor<'db, 'a> {
    db: &'db dyn HirAnalysisDb,
    inner: name_resolver::NameResolver<'db, 'a>,
    diags: Vec<diagnostics::NameResDiag>,
    item_stack: Vec<ItemKind>,
    path_ctxt: Vec<ExpectedPathKind>,

    /// The set of scopes that have already been conflicted to avoid duplicate
    /// diagnostics.
    already_conflicted: FxHashSet<ScopeId>,
}

impl<'db, 'a> EarlyPathVisitor<'db, 'a> {
    fn new(db: &'db dyn HirAnalysisDb, importer: &'a DefaultImporter) -> Self {
        let resolver = name_resolver::NameResolver::new(db, importer);
        Self {
            db: db.as_hir_analysis_db(),
            inner: resolver,
            diags: Vec::new(),
            item_stack: Vec::new(),
            path_ctxt: Vec::new(),
            already_conflicted: FxHashSet::default(),
        }
    }

    fn verify_path(
        &mut self,
        path: PathId,
        scope: ScopeId,
        span: LazyPathSpan,
        bucket: NameResBucket,
    ) {
        let path_kind = self.path_ctxt.last().unwrap();
        let last_seg_idx = path.len(self.db.as_hir_db()) - 1;
        let last_seg_ident = *path.segments(self.db.as_hir_db())[last_seg_idx].unwrap();
        let span = span.segment(last_seg_idx).into();

        if bucket.is_empty() {
            let Err(err) = bucket.pick(path_kind.domain()) else {
                unreachable!()
            };

            match err {
                NameResolutionError::NotFound => {
                    self.diags
                        .push(NameResDiag::not_found(span, last_seg_ident));
                }
                NameResolutionError::Ambiguous(cands) => {
                    self.diags.push(NameResDiag::ambiguous(
                        self.db,
                        span,
                        last_seg_ident,
                        cands.clone(),
                    ));
                }
                _ => {}
            };

            return;
        }

        match path_kind.pick(bucket) {
            // The path exists and belongs to the expected kind.
            Either::Left(res) => {
                if !res.is_visible(self.db, scope) {
                    self.diags.push(NameResDiag::invisible(
                        span,
                        last_seg_ident,
                        res.derived_from(self.db),
                    ));
                }
            }

            // The path exists but doesn't belong to the expected kind.
            Either::Right(res) => match path_kind {
                ExpectedPathKind::Type => {
                    self.diags
                        .push(NameResDiag::ExpectedType(span, last_seg_ident, res));
                }

                ExpectedPathKind::Trait => {
                    self.diags
                        .push(NameResDiag::ExpectedTrait(span, last_seg_ident, res));
                }

                ExpectedPathKind::Value => {
                    self.diags
                        .push(NameResDiag::ExpectedValue(span, last_seg_ident, res));
                }
            },
        }
    }

    fn check_conflict(&mut self, scope: ScopeId) {
        if !self.already_conflicted.insert(scope) {
            return;
        }

        let Some(query) = self.make_query_for_conflict_check(scope) else {
            return;
        };

        let domain = NameDomain::from_scope(self.db, scope);
        let binding = self.inner.resolve_query(query);
        match binding.pick(domain) {
            Ok(_) => {}

            Err(NameResolutionError::Ambiguous(cands)) => {
                let conflicted_span = cands
                    .iter()
                    .filter_map(|res| {
                        let conflicted_scope = res.scope()?;
                        self.already_conflicted.insert(conflicted_scope);
                        conflicted_scope.name_span(self.db.as_hir_db())
                    })
                    .collect();

                let diag = diagnostics::NameResDiag::conflict(
                    scope.name(self.db.as_hir_db()).unwrap(),
                    conflicted_span,
                );
                self.diags.push(diag);
            }

            Err(_) => unreachable!(),
        };
    }

    fn make_query_for_conflict_check(&self, scope: ScopeId) -> Option<NameQuery> {
        let name = scope.name(self.db.as_hir_db())?;
        let mut directive = QueryDirective::new();
        directive.disallow_lex().disallow_glob().disallow_external();

        let parent_scope = scope.parent(self.db.as_hir_db())?;
        Some(NameQuery::with_directive(name, parent_scope, directive))
    }
}

impl<'db, 'a> Visitor for EarlyPathVisitor<'db, 'a> {
    fn visit_item(&mut self, ctxt: &mut VisitorCtxt<'_, LazyItemSpan>, item: ItemKind) {
        // We don't need to check use statements for conflicts because they are
        // already checked in import resolution.
        if matches!(item, ItemKind::Use(_)) {
            return;
        }

        // We don't need to check impl blocks for conflicts because they
        // needs ingot granularity analysis, the conflict checks for them is done by the
        // `ImplCollector`.
        if !matches!(item, ItemKind::Impl(_)) {
            let scope = ScopeId::from_item(item);
            self.check_conflict(scope);
        }

        self.item_stack.push(item);
        if matches!(item, ItemKind::Body(_)) {
            self.path_ctxt.push(ExpectedPathKind::Value);
        } else {
            self.path_ctxt.push(ExpectedPathKind::Type);
        }

        walk_item(self, ctxt, item);

        self.item_stack.pop();
        self.path_ctxt.pop();
    }

    fn visit_trait_ref(
        &mut self,
        ctxt: &mut VisitorCtxt<'_, LazyTraitRefSpan>,
        trait_ref: TraitRefId,
    ) {
        self.path_ctxt.push(ExpectedPathKind::Trait);
        walk_trait_ref(self, ctxt, trait_ref);
        self.path_ctxt.pop();
    }

    fn visit_field_def(
        &mut self,
        ctxt: &mut VisitorCtxt<'_, LazyFieldDefSpan>,
        field: &hir::hir_def::FieldDef,
    ) {
        let scope = ctxt.scope();
        self.check_conflict(scope);
        walk_field_def(self, ctxt, field);
    }

    fn visit_variant_def(
        &mut self,
        ctxt: &mut VisitorCtxt<'_, LazyVariantDefSpan>,
        variant: &hir::hir_def::VariantDef,
    ) {
        let scope = ctxt.scope();
        self.check_conflict(scope);
        walk_variant_def(self, ctxt, variant);
    }

    fn visit_generic_param(
        &mut self,
        ctxt: &mut VisitorCtxt<'_, LazyGenericParamSpan>,
        param: &hir::hir_def::GenericParam,
    ) {
        let scope = ctxt.scope();
        self.check_conflict(scope);
        walk_generic_param(self, ctxt, param);
    }

    fn visit_generic_arg_list(
        &mut self,
        ctxt: &mut VisitorCtxt<'_, LazyGenericArgListSpan>,
        args: GenericArgListId,
    ) {
        self.path_ctxt.push(ExpectedPathKind::Type);
        walk_generic_arg_list(self, ctxt, args);
        self.path_ctxt.pop();
    }

    fn visit_ty(&mut self, ctxt: &mut VisitorCtxt<'_, LazyTySpan>, ty: TypeId) {
        self.path_ctxt.push(ExpectedPathKind::Type);
        walk_ty(self, ctxt, ty);
        self.path_ctxt.pop();
    }

    // We don't need to run path analysis on patterns, statements and expressions in
    // early path resolution.
    fn visit_pat(&mut self, _: &mut VisitorCtxt<'_, LazyPatSpan>, _: PatId, _: &Pat) {}

    fn visit_expr(
        &mut self,
        ctxt: &mut VisitorCtxt<'_, LazyExprSpan>,
        expr: ExprId,
        expr_data: &Expr,
    ) {
        if matches!(expr_data, Expr::Block(_)) {
            walk_expr(self, ctxt, expr)
        }
    }

    fn visit_path(&mut self, ctxt: &mut VisitorCtxt<'_, LazyPathSpan>, path: PathId) {
        let scope = ctxt.scope();
        let dummy_cache_store = ResolvedQueryCacheStore::no_cache();

        let mut resolver = EarlyPathResolver::new(self.db, &mut self.inner, &dummy_cache_store);
        let resolved_path = match resolver.resolve_path(path, scope) {
            Ok(bucket) => bucket,

            Err(err) => {
                let failed_at = err.failed_at;
                let span = ctxt.span().unwrap().segment(failed_at);
                let ident = path.segments(self.db.as_hir_db())[failed_at];

                let diag = match err.kind {
                    NameResolutionError::NotFound => {
                        NameResDiag::not_found(span.into(), *ident.unwrap())
                    }

                    NameResolutionError::Invalid => {
                        return;
                    }

                    NameResolutionError::Invisible(_) => {
                        unreachable!("`EarlyPathResolver doesn't check visibility");
                    }

                    NameResolutionError::Ambiguous(cands) => {
                        NameResDiag::ambiguous(self.db, span.into(), *ident.unwrap(), cands)
                    }

                    NameResolutionError::InvalidPathSegment(res) => {
                        NameResDiag::invalid_use_path_segment(
                            self.db,
                            span.into(),
                            *ident.unwrap(),
                            res,
                        )
                    }

                    NameResolutionError::Conflict(name, spans) => {
                        NameResDiag::Conflict(name, spans)
                    }
                };

                self.diags.push(diag);
                return;
            }
        };

        if let Some((idx, res)) = resolved_path.find_invisible_segment(self.db) {
            let span = ctxt.span().unwrap().segment(idx);
            let ident = path.segments(self.db.as_hir_db())[idx].unwrap();
            let diag = NameResDiag::invisible(span.into(), *ident, res.derived_from(self.db));
            self.diags.push(diag);
            return;
        }

        let EarlyResolvedPath::Full(bucket) = resolved_path.resolved else {
            return;
        };
        self.verify_path(path, scope, ctxt.span().unwrap(), bucket);
    }
}

#[derive(Debug, Clone, Copy)]
enum ExpectedPathKind {
    Type,
    Trait,
    Value,
}

impl ExpectedPathKind {
    fn domain(self) -> NameDomain {
        match self {
            ExpectedPathKind::Type => NameDomain::Type,
            ExpectedPathKind::Trait => NameDomain::Type,
            ExpectedPathKind::Value => NameDomain::Value,
        }
    }

    fn pick(self, bucket: NameResBucket) -> Either<NameRes, NameRes> {
        debug_assert!(!bucket.is_empty());

        let res = match bucket.pick(self.domain()).as_ref().ok() {
            Some(res) => res.clone(),
            None => {
                return Either::Right(bucket.into_iter().find_map(|res| res.ok()).unwrap());
            }
        };

        match self {
            Self::Type if !res.is_type() => Either::Right(res),
            Self::Trait if !res.is_trait() => Either::Right(res),
            Self::Value if !res.is_value() => Either::Right(res),
            _ => Either::Left(res),
        }
    }
}
