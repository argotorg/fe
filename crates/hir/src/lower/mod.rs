use fe_parser2::{ast, SyntaxToken};
use num_bigint::BigUint;
use num_traits::Num;

use crate::{
    hir_def::{IdentId, IntegerId, LitKind, MaybeInvalid, StringId},
    HirDb,
};

mod attr;
mod body;
mod expr;
mod item;
mod params;
mod pat;
mod path;
mod stmt;
mod types;
mod use_tree;

impl IdentId {
    fn from_token(db: &dyn HirDb, token: SyntaxToken) -> Self {
        Self::new(db, token.text().to_string())
    }

    fn maybe_from_token(db: &dyn HirDb, token: Option<SyntaxToken>) -> MaybeInvalid<Self> {
        token.map(|token| Self::from_token(db, token)).into()
    }
}

impl LitKind {
    pub(super) fn from_ast(db: &dyn HirDb, ast: ast::Lit) -> Self {
        match ast.kind() {
            ast::LitKind::Int(int) => Self::Int(IntegerId::from_ast(db, int)),
            ast::LitKind::String(string) => {
                let text = string.token().text();
                Self::String(StringId::new(db, text[1..text.len() - 1].to_string()))
            }
            ast::LitKind::Bool(bool) => match bool.token().text() {
                "true" => Self::Bool(true),
                "false" => Self::Bool(false),
                _ => unreachable!(),
            },
        }
    }
}

impl IntegerId {
    pub(super) fn from_ast(db: &dyn HirDb, ast: ast::LitInt) -> Self {
        let text = ast.token().text();
        // Parser ensures that the text is valid pair with a radix and a number.
        if text.len() < 2 {
            return Self::new(db, BigUint::from_str_radix(&text, 10).unwrap());
        }

        let int = match &text[0..2] {
            "0x" | "0X" => BigUint::from_str_radix(&text[2..], 16).unwrap(),
            "0o" | "0O" => BigUint::from_str_radix(&text[2..], 8).unwrap(),
            "0b" | "0B" => BigUint::from_str_radix(&text[2..], 2).unwrap(),
            _ => BigUint::from_str_radix(&text, 10).unwrap(),
        };

        Self::new(db, int)
    }
}
