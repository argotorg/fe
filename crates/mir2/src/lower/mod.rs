// pub mod function;
// pub mod types;

// mod pattern_match;

pub fn lower() {}
