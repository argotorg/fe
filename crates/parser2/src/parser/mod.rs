use fxhash::FxHashSet;

use crate::{ParseError, SyntaxKind, TextRange};

use self::token_stream::{BackTrackableTokenStream, SyntaxToken, TokenStream};

pub mod token_stream;

/// Parser to build a rowan syntax tree.
pub struct Parser<S: TokenStream> {
    /// Token stream to parse.
    stream: BackTrackableTokenStream<S>,

    builder: rowan::GreenNodeBuilder<'static>,
    scopes: Vec<Box<dyn ParsingScope>>,
    errors: Vec<ParseError>,

    /// The checkpoint where the scope/branch is wrapped up later by another
    /// scope/branch when the `wrap_scope_with` method.
    check_point: Option<(rowan::Checkpoint, usize)>,

    current_pos: rowan::TextSize,
}

impl<S: TokenStream> Parser<S> {
    /// Returns the current token of the parser.
    pub fn current_token(&mut self) -> Option<&S::Token> {
        self.stream.peek()
    }

    /// Enters the scope and set the `scope` to the current scope.
    /// If `is_checkpoint` is true, the current scope/branch is wrapped up by
    /// another scope/branch later when the [`wrap_scope_with`] method is
    /// called.
    pub fn enter(&mut self, scope: Box<dyn ParsingScope>, is_checkpoint: bool) {
        if is_checkpoint {
            self.check_point = Some((self.builder.checkpoint(), self.scopes.len()));
        }

        self.builder.start_node(scope.syntax_kind().into());
        self.scopes.push(scope);
    }

    /// Enters the errors scope and add the `msg` to the error list.
    pub fn enter_with_error(&mut self, msg: &str) {
        let start = self.current_pos;
        let end = if let Some(current_token) = self.current_token() {
            start + current_token.text_size()
        } else {
            start
        };
        let range = TextRange::new(start, end);

        self.errors.push(ParseError {
            range,
            msg: msg.to_string(),
        });
        self.scopes.push(Box::new(ErrorScope()));
    }

    /// Leaves the current scope/branch.
    pub fn leave(&mut self) {
        self.scopes.pop();
        self.builder.finish_node();
    }

    /// Wrap up the marked scope/branch with another scope/branch and set the
    /// `scope` to the current scope.
    pub fn wrap_scope_with(&mut self, scope: Box<dyn ParsingScope>) {
        debug_assert!(self.check_point.is_some(), "No checkpoint");
        let check_point = self.check_point.take().unwrap();
        let syntax_kind = scope.syntax_kind();

        self.scopes.truncate(check_point.1);
        self.scopes.push(scope);

        self.builder
            .start_node_at(check_point.0, syntax_kind.into());
    }

    /// Bumps the current token and adds it to the current branch.
    pub fn bump(&mut self) {
        let tok = self.stream.next().unwrap();
        self.current_pos += rowan::TextSize::of(tok.text());
        self.builder.token(tok.syntax_kind().into(), tok.text());
    }

    /// Bumps consecutive trivia tokens.
    pub fn bump_trivias(&mut self) {
        while let Some(tok) = self.current_token() {
            let kind = tok.syntax_kind();
            if kind.is_trivia() {
                self.bump();
            } else {
                break;
            }
        }
    }

    /// Bump consecutive newlines.
    pub fn bump_newlines(&mut self) {
        while let Some(tok) = self.current_token() {
            if tok.syntax_kind() == SyntaxKind::Newline {
                self.bump();
            } else {
                break;
            }
        }
    }

    /// Proceeds the parser to the recovery token of the current scope. Then
    /// leave the current branch/scope.
    pub fn recovery(&mut self) {
        let mut scope_index = self.scopes.len() - 1;
        // Finds the nearest scope that has its own recovery set.
        loop {
            if self.scopes[scope_index].recovery_method() != &RecoveryMethod::Inheritance
                || scope_index == 0
            {
                break;
            } else {
                scope_index -= 1;
            }
        }

        while let Some(tok) = self.stream.peek() {
            let syntax_kind = tok.syntax_kind();
            if self.scopes[scope_index]
                .recovery_method()
                .contains(syntax_kind)
            {
                break;
            } else {
                self.bump();
            }
        }

        self.leave();
    }
}

/// The current scope of parsing.
pub trait ParsingScope {
    /// Returns the recovery method of the current scope.
    fn recovery_method(&self) -> &RecoveryMethod;

    fn syntax_kind(&self) -> SyntaxKind;
}

pub struct ErrorScope();

impl ParsingScope for ErrorScope {
    fn recovery_method(&self) -> &RecoveryMethod {
        &RecoveryMethod::Inheritance
    }

    fn syntax_kind(&self) -> SyntaxKind {
        SyntaxKind::Error
    }
}

/// Represents the recovery method of the current scope.
#[derive(PartialEq, Eq)]
pub enum RecoveryMethod {
    /// Uses the recovery method of the parent scope.
    Inheritance,

    /// The scope has its own recovery set.
    RecoverySet(FxHashSet<SyntaxKind>),
}

impl RecoveryMethod {
    /// Returns `true` if the recovery set contains the given syntax kind.
    fn contains(&self, syntax_kind: SyntaxKind) -> bool {
        match self {
            RecoveryMethod::Inheritance => false,
            RecoveryMethod::RecoverySet(set) => set.contains(&syntax_kind),
        }
    }
}

trait TextSize {
    fn text_size(&self) -> rowan::TextSize;
}

impl<T> TextSize for T
where
    T: SyntaxToken,
{
    fn text_size(&self) -> rowan::TextSize {
        rowan::TextSize::of(self.text())
    }
}
