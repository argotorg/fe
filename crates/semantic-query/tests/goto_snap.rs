use common::InputDb;
use dir_test::{dir_test, Fixture};
use driver::DriverDataBase;
use fe_semantic_query::SemanticIndex;
use hir::lower::{map_file_to_mod, parse_file_impl};
use hir_analysis::name_resolution::{resolve_with_policy, DomainPreference, PathResErrorKind};
use parser::SyntaxNode;
use test_utils::snap_test;
use url::Url;

fn collect_positions(root: &SyntaxNode) -> Vec<parser::TextSize> {
    use parser::{ast, ast::prelude::AstNode, SyntaxKind};
    fn walk(node: &SyntaxNode, positions: &mut Vec<parser::TextSize>) {
        match node.kind() {
            SyntaxKind::Ident => positions.push(node.text_range().start()),
            SyntaxKind::Path => {
                if let Some(path) = ast::Path::cast(node.clone()) {
                    for segment in path.segments() {
                        if let Some(ident) = segment.ident() {
                            positions.push(ident.text_range().start());
                        }
                    }
                }
            }
            SyntaxKind::PathType => {
                if let Some(pt) = ast::PathType::cast(node.clone()) {
                    if let Some(path) = pt.path() {
                        for segment in path.segments() {
                            if let Some(ident) = segment.ident() {
                                positions.push(ident.text_range().start());
                            }
                        }
                    }
                }
            }
            SyntaxKind::FieldExpr => {
                if let Some(fe) = ast::FieldExpr::cast(node.clone()) {
                    if let Some(tok) = fe.field_name() {
                        positions.push(tok.text_range().start());
                    }
                }
            }
            SyntaxKind::UsePath => {
                if let Some(up) = ast::UsePath::cast(node.clone()) {
                    for seg in up.into_iter() {
                        if let Some(tok) = seg.ident() { positions.push(tok.text_range().start()); }
                    }
                }
            }
            _ => {}
        }
        for child in node.children() {
            walk(&child, positions);
        }
    }
    let mut out = Vec::new();
    walk(root, &mut out);
    out.sort();
    out.dedup();
    out
}

fn line_col_from_cursor(cursor: parser::TextSize, s: &str) -> (usize, usize) {
    let mut line = 0usize;
    let mut col = 0usize;
    for (i, ch) in s.chars().enumerate() {
        if i == Into::<usize>::into(cursor) {
            return (line, col);
        }
        if ch == '\n' { line += 1; col = 0; } else { col += 1; }
    }
    (line, col)
}

fn format_snapshot(content: &str, lines: &[String]) -> String {
    let header = content
        .lines()
        .enumerate()
        .map(|(i, l)| format!("{i:?}: {l}"))
        .collect::<Vec<_>>()
        .join("\n");
    let body = lines.join("\n");
    format!("{header}\n---\n{body}")
}

#[dir_test(dir: "$CARGO_MANIFEST_DIR/test_files/goto", glob: "*.fe")]
fn test_goto_snapshot(fixture: Fixture<&str>) {
    if fixture.path().ends_with("use_paths.fe") {
        return;
    }
    let mut db = DriverDataBase::default();
    let file = db.workspace().touch(
        &mut db,
        Url::from_file_path(fixture.path()).unwrap(),
        Some(fixture.content().to_string()),
    );
    let top_mod = map_file_to_mod(&db, file);

    // Parse and collect identifier/path-segment positions
    let green = parse_file_impl(&db, top_mod);
    let root = SyntaxNode::new_root(green);
    let positions = collect_positions(&root);

    let mut lines = Vec::new();
    for cursor in positions {
        // Use SemanticIndex to pick segment subpath and check def candidates count
        let count = SemanticIndex::goto_candidates_at_cursor(&db, &db, top_mod, cursor).len();

        // Resolve pretty path(s) for readability
        let pretty = if let Some((path, scope, seg_idx, _)) = SemanticIndex::at_cursor(&db, top_mod, cursor) {
            let seg_path = path.segment(&db, seg_idx).unwrap_or(path);
            match resolve_with_policy(&db, seg_path, scope, hir_analysis::ty::trait_resolution::PredicateListId::empty_list(&db), DomainPreference::Either) {
                Ok(res) => vec![res.pretty_path(&db).unwrap_or("<unknown>".into())],
                Err(err) => match err.kind {
                    PathResErrorKind::NotFound { bucket, .. } => bucket.iter_ok().filter_map(|nr| nr.pretty_path(&db)).collect(),
                    PathResErrorKind::Ambiguous(vec) => vec.into_iter().filter_map(|nr| nr.pretty_path(&db)).collect(),
                    _ => vec![],
                }
            }
        } else { vec![] };

        if !pretty.is_empty() || count > 0 {
            let (line, col) = line_col_from_cursor(cursor, fixture.content());
            let joined = pretty.join("\n");
            lines.push(format!("cursor position ({line}, {col}), {count} defs -> {joined}"));
        }
    }

    let snapshot = format_snapshot(fixture.content(), &lines);
    snap_test!(snapshot, fixture.path());
}
