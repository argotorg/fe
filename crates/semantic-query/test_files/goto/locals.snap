---
source: crates/semantic-query/tests/goto_snap.rs
assertion_line: 130
expression: snapshot
input_file: test_files/goto/locals.fe
---
0: fn test_locals(x: i32, y: i32) -> i32 {
1:   let a = x
2:   let x = a + y
3:   x
4: }
5: 
---
cursor position (0, 18), 0 defs -> i32
cursor position (0, 26), 0 defs -> i32
cursor position (0, 34), 0 defs -> i32
cursor position (1, 6), 1 defs -> 
cursor position (1, 10), 1 defs -> locals::test_locals::x
cursor position (2, 6), 1 defs -> locals::test_locals::x
cursor position (2, 10), 1 defs -> 
cursor position (2, 14), 1 defs -> locals::test_locals::y
cursor position (3, 2), 1 defs -> locals::test_locals::x
